docs/docs/schema-customization.md

```diff
@@ -0,0 +1,591 @@
+# Customizing the GraphQL Schema
+
+One of Gatsby's main strengths is the ability to query data from a variety of
+sources in a uniform way with GraphQL. For this to work, a GraphQL Schema must
+be generated that defines the shape of the data.
+
+Gatsby is able to automatically infer a GraphQL Schema from your data, and in
+many cases this is really all you need. There are however situations when you
+either want to explicitly define the data shape, or add custom functionality to
+the query layer - this is what Gatsby's Schema Customization API provides.
+
+The following guide walks through a basic example to showcase the API.
+
+# Explicitly defining data types
+
+Our example project is a blog that gets its data from local Markdown files which
+provide the post contents, as well as author information in JSON format. We also
+have occasional guest contributors whose info we keep in a separate JSON file.
+
+```markdown:title=src/data/post1.md
+---
+title: Sample Post
+publishedAt: 2019-04-01
+author: jane@example.com
+tags:
+  - wow
+---
+
+# Heading
+
+Text
+```
+
+```json:title=src/data/author.json
+[
+  {
+    "name": "Doe",
+    "firstName": "Jane",
+    "email": "jane@example.com",
+    "joinedAt": "2018-01-01"
+  }
+]
+```
+
+```json:title=src/data/contributor.json
+[
+  {
+    "name": "Doe",
+    "firstName": "Zoe",
+    "email": "zoe@example.com",
+    "receivedSwag": true
+  }
+]
+```
+
+To be able to query the contents of these files with GraphQL, we need to first
+load them into Gatsby's internal data store. This is what source and transformer
+plugin accomplish - in this case `gatsby-source-filesystem` and
+`gatsby-transformer-remark` plus `gatsby-transformer-json`. Every markdown post
+file is hereby transformed into a "node" object in the internal data store with
+a unique `id` and a type `MarkdownRemark`. Similarly, an author will be
+represented by a node object of type `AuthorJson`, and contributor info will be
+transformed into node objects of type `ContributorJson`.
+
+## The Node interface
+
+This data structure is represented in Gatsby's GraphQL schema with the `Node`
+interface, which describes the set of fields common to node objects created by
+source and transformer plugins (`id`, `parent`, `children`, as well as a couple
+of `internal` fields like `type`). In GraphQL Schema Definition Language (SDL),
+it looks like this:
+
+```graphql
+interface Node {
+  id: ID!
+  parent: Node!
+  children: [Node!]!
+  internal: Internal!
+}
+
+type Internal {
+  type: String!
+}
+```
+
+Types created by source and transformer plugins implement this interface. For
+example, the node type created by `gatsby-transformer-json` for `authors.json`
+will be represented in the GraphQL schema as:
+
+```graphql
+type AuthorJson implements Node {
+  id: ID!
+  parent: Node!
+  children: [Node!]!
+  internal: Internal!
+  name: String
+  firstName: String
+  email: String
+  joinedAt: Date
+}
+```
+
```

Comment:
maybe add a section how i find the generated schemas?

Reply:
I ask this because I tried to get India and the only way was via the docs section in the iql explorer. But there was no place where I get exactly the scheme definition (for copy paste etc) which you posted in the code snippets

Reply:
the GraphQL Playground also has a `Schema` tab, you can get it if you start with
```shell
GATSBY_GRAPHQL_IDE=playground gatsby develop
```
is that what you mean?

Reply:
ah - yes :)

it would be helpful to note this option and the schema tab to see the resulting schema better.


or make a note with a link to this:
https://www.gatsbyjs.org/docs/why-gatsby-uses-graphql/
https://www.gatsbyjs.org/docs/using-graphql-playground/